src/RevCPU.cc

```diff
@@ -2660,95 +2663,92 @@ void RevCPU::UpdateThreadAssignments(uint32_t ProcID){
-          break;
-        case ThreadState::BLOCKED:
-          // This thread is blocked (currently only caused by a rev_pthread_join)
-          // We need to:
-          // 1. Check if the thread it is waiting on has already completed
-          // 2. If it has... Thread can resume execution
-          // 3. If not, thread remains blocked
-          //    3a. Move its ThreadID to the BlockedThreads list
-          //    3b. Remove it from the AssignedThreads lis
-          output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 "on Core %" PRIu32 " is BLOCKED\n", Thread->GetThreadID(), ProcID);
-          // -- 1.
-          if( ThreadCanProceed(Thread->GetThreadID()) ){
-            // -- 2.
-            output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has already completed so it can proceed\n",
-                           Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
-            // Continue executing thread on same Core
-            Thread->SetState(ThreadState::RUNNING);
-          }
-          else { // -- 3.
-            output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has not yet completed so it remains blocked\n",
-                           Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
-            Thread->SetState(ThreadState::BLOCKED);
-            // -- 3a.
-            BlockedThreads.emplace(Thread->GetThreadID());
-            // -- 3b.
-            AssignedThreads.at(ProcID).erase(AssignedThreads.at(ProcID).begin()+HartID);
-          }
-          break;
-        case ThreadState::START: // Should never happen
-          output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is assigned but is in START state... This is a bug\n",
-                       Thread->GetThreadID(), ProcID);
-          break;
-        case ThreadState::RUNNING:
-          output.verbose(CALL_INFO, 11, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is RUNNING\n", Thread->GetThreadID(), ProcID);
-          break;
-        case ThreadState::READY:
-          // If this happens we are not setting state when assigning thread somewhere
-          output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is assigned but is in START state... This is a bug\n",
-                       Thread->GetThreadID(), ProcID);
-          break;
-        default: // Should DEFINITELY never happen
-          output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug\n",
-                       Thread->GetThreadID(), ProcID);
-          break;
-        }
-        // State change has been handled, so clear the bit
-        Changes[HartID] = false;
+  // NOTE: At this point we handle EVERY thread that changed state
+  while( !Procs[ProcID]->GetThreadsThatChangedState().empty() ){
+    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
```

Comment:
FYI:
`auto& Thread` means a reference to an lvalue (non-temporary which has an address) which might be `const`

`auto Thread` means a local copy of a value

`auto&& Thread` means a reference to an lvalue or rvalue, depending on the RHS, which might be `const`. It is the "universal reference" which can bind to anything. If you don't know anything about the RHS, then `auto&&` is the most universal.

Reply:
I definitely don't want a local copy. If I know the pointer isn't `const` is the single `&` sufficient? Or should I use the `&&`?

Reply:
If you know the RHS is an lvalue, or you want to enforce it as an lvalue, then `auto&` is correct. `auto&&` is only needed in variable contexts where the RHS can be either a lvalue or rvalue and might change. This lvalue/rvalue distinction is independent of `const`ness, since `auto` reference can deduce `const`.

For example, `auto&& [a, b, c]` is a structured binding with a universal reference which will work when the RHS has a tuple-like mixture of lvalues and rvalues.